### ISSUE TITLE
[Example] Face landmark detection

### ISSUE THREAD

**Issue by Wenzhao-Xiang** [intel/webml-polyfill#373]
I did some investigation for face landmark detection. It seems not so good to implement an end-to-end model, since it's difficult for one model to detect face feature from a whole image.

So, plan to separate this task to two steps with two models: 
1. Face detection for input image
2. For each face, detect its key points

**Comment by huningxin** [comment]
There are some potential references:
1. [DeepFace](https://github.com/RiweiChen/DeepFace)
2. [FaceNet](https://github.com/davidsandberg/facenet)
3. Face detection and landmarks detection models of [OpenVINO model zoo](https://software.intel.com/en-us/openvino-toolkit/documentation/pretrained-models)
4. Face detection and recognition models of [ONNX model zoo](https://github.com/onnx/models#face-detection-and-recognition-)
5. [OpenFace](https://github.com/cmusatyalab/openface) used by [OpenCV.js dnn example](https://docs.opencv.org/master/d5/d86/tutorial_dnn_javascript.html)

**Comment by Wenzhao-Xiang** [comment]
Nice references! Thanks!

**Comment by Wenzhao-Xiang** [comment]
For face detection, there are three popular models now: 
1. SSD 
2. YOLO
3. MTCNN
The first two models are migrated from object detection with only one class "face". And the third model is used for face detection and alignment, which can output both detected face and five key points.

For worrying of performance on web, first I will have a try with Tiny-YOLO-v2 for face detection.

**Comment by Wenzhao-Xiang** [comment]
1. I implemented tiny-yolo-v2 for face detection 
Reference: [YoloKerasFaceDetection](https://github.com/abars/YoloKerasFaceDetection)

2. This model is  migrated from Tiny-YOLO-V2 object detection model. The original model is like this. 
![image](https://user-images.githubusercontent.com/40754037/50445692-541e9b80-094b-11e9-82b2-3bc78a769e69.png)
The output of face detection model is reduced to [13,13,30], because there is only one class `face` in label list.(The original model's output is [13,13,125] with 20 classes)

3. Tiny-YOLO-V2 uses new operation [LeakyRelu](https://www.tensorflow.org/api_docs/python/tf/nn/leaky_relu) . But when convert keras model to tflite model, LeakyRelu is replaced by MUL+[Maximum](https://www.tensorflow.org/api_docs/python/tf/math/maximum), like this:
![image](https://user-images.githubusercontent.com/40754037/50446070-cee8b600-094d-11e9-9d58-913f05e42826.png)
So new op we need is `Maximum`.

4. A simple image demo with WebGL backend. 
input size [416, 416, 3] => output size [13, 13, 30]
GPU: 1080TI

Single face:
![image](https://user-images.githubusercontent.com/40754037/50446467-035d7180-0950-11e9-9bb2-6c3ab9554784.png)


Multiply face:
![image](https://user-images.githubusercontent.com/40754037/50446377-abbf0600-094f-11e9-94ac-9de871890b78.png)

Tiny-yolo-V2 is fast but with low accuracy when multiply face. I will try other models later.

**Comment by Wenzhao-Xiang** [comment]
Visit https://wenzhao-xiang.github.io/intel/webml-polyfill/ for face detection demo.

**Comment by Wenzhao-Xiang** [comment]
A face detection video demo, about **29~30 FPS** with **1080Ti**
![image](https://user-images.githubusercontent.com/40754037/50505609-b09acb80-0aaf-11e9-80f7-fcc1d718bc3c.png)

**Comment by Wenzhao-Xiang** [comment]
I investigated MTCNN last week, but it seems that MTCNN isn't suitable for our API. 
MTCNN models have three stages.
1. Stage 1: Proposal Net, provides a large number of candidate windows, computes the bounding box regression vector and reduces the window with NMS.
2. Stage 2: Refining Net, rejects some non-face windows, calculates bounding box regression vectors and reduces the window with NMS.
3. Stage 3: Outputting Net, output 5 landmark points on the face.

The first stage need different scaled input images, which means an uncertain input tensor for the first network. However, our API need all of the tensors to be certain when initializing, conflicting with MTCNN.

**Comment by huningxin** [comment]
> conflicting with MTCNN.

Thanks for the investigation. I am wondering how MTCNN implements by using DL frameworks. Do you have any ideas?

**Comment by Wenzhao-Xiang** [comment]
After use SSD model for face detector, landmark detection result work better now!
The face detection models are SSD_MobilenetV1, SSD_MobilenetV2, SSDLite_MobilenetV2 and Tiny_YoloV2. And face alignment model is DAN(Deep Alignment Network, [here](https://arxiv.org/pdf/1706.01789.pdf) is the paper).

I have pushed my code to my github io, please go to 
https://wenzhao-xiang.github.io/intel/webml-examples/index.html
for a simple face detection and alignment image/camera demo.

Single face:
![image](https://user-images.githubusercontent.com/40754037/51443066-c8681700-1d1e-11e9-9c1b-201149ef54a9.png)

Multiple face:
![image](https://user-images.githubusercontent.com/40754037/51443132-38769d00-1d1f-11e9-88f9-89d67cf73e57.png)
